src/Compilers/CSharp/Portable/BoundTree/Expression.cs

```diff
@@ -170,9 +175,15 @@ private static IArgument DeriveArgument(int parameterIndex, int argumentIndex, I
                         RefKind refMode = !argumentRefKinds.IsDefaultOrEmpty ? argumentRefKinds[argumentIndex] : RefKind.None;
                         return
                             refMode == RefKind.None
-                            ? ((argumentIndex >= parameters.Length - 1 && parameters.Length > 0 && parameters[parameters.Length - 1].IsParams)
-                                ? (IArgument)new Argument(ArgumentKind.ParamArray, parameters[parameters.Length - 1], CreateParamArray(parameters[parameters.Length - 1], boundArguments, argumentIndex))
-                                : new SimpleArgument(parameters[parameterIndex], argument))
+                            ? (argumentIndex >= parameters.Length - 1 &&
+                               parameters.Length > 0 &&
+                               parameters[parameters.Length - 1].IsParams &&
+                               // An argument that is an array of the appropriate type is not a params argument.
+                               (boundArguments.Length > argumentIndex + 1 ||
+                                argument.Type.TypeKind != TypeKind.Array ||
+                                !argument.Type.Equals(parameters[parameters.Length - 1].Type, true))
+                               ? (IArgument)new Argument(ArgumentKind.ParamArray, parameters[parameters.Length - 1], CreateParamArray(parameters[parameters.Length - 1], boundArguments, argumentIndex))
+                               : new SimpleArgument(parameters[parameterIndex], argument))
```

Comment:
I feel like this conditional shoud be broken up into `if` blocks.  In the current form it's very hard to follow.